README.md

```diff
@@ -0,0 +1,140 @@
+    ),
+    "prompt_no_input": (
+        "Below is an instruction that describes a task. "
+        "Write a response that appropriately completes the request.\n\n"
+        "### Instruction:\n{instruction}\n\n### Response:"
+    ),
+}
+
+def format_alpaca_fn(example):
+    prompt_input, prompt_no_input = PROMPT_DICT['prompt_input'], PROMPT_DICT['prompt_no_input']
+    output = prompt_input.format_map(example) if example.get("input", "") != "" else prompt_no_input.format_map(example)
+    output = f"{output} {example['output']}"
+    return {"output": output}
+
+ds = datasets.load_dataset('json', data_files='./stanford_alpaca/alpaca_data.json')
+
+alpaca_ds = ds['train'].map(format_alpaca_fn, remove_columns=['instruction', 'input'])
+alpaca_ds.to_json("sft_alpaca_data.json")
+```
+
+The `response template` corresponding to the above dataset and the `Llama` tokenizer is: `\n### Response:"`.
+
+The same way can be applied to any dataset, with more info can be found [here](https://huggingface.co/docs/trl/main/en/sft_trainer#format-your-input-prompts).
+
+
+## Supported Models
+
+Current supported and tested models are `Llama2` (7 and 13B configurations have been tested) and `GPTBigCode`.
+
+## Training
+
+### Single GPU
+```python
+# if you want to use one GPU on multi-gpu machine
+export CUDA_VISIBLE_DEVICES=0
+
+python tuning/sft_trainer.py  \
+--model_name_or_path $MODEL_PATH  \
+--data_path $DATA_PATH  \
+--output_dir $OUTPUT_PATH  \
+--num_train_epochs 5  \
+--per_device_train_batch_size 4  \
+--per_device_eval_batch_size 4  \
+--gradient_accumulation_steps 4  \
+--evaluation_strategy "no"  \
+--save_strategy "epoch"  \
+--learning_rate 1e-5  \
+--weight_decay 0.  \
+--warmup_ratio 0.03  \
+--lr_scheduler_type "cosine"  \
+--logging_steps 1  \
+--include_tokens_per_second  \
+--packing False  \
+--response_template "\n### Response:"  \
+--dataset_text_field "output" 
+
+```
+
+### Multiple GPUs with FSDP
+```
+torchrun \
+--nnodes=1 \
+--nproc_per_node=8 \ 
+--master_port=1234 \
+tuning/sft_trainer.py \
+--model_name_or_path $MODEL_PATH \
+--data_path $DATA_PATH \
+--bf16 True \
+--output_dir $OUTPUT_PATH \
+--num_train_epochs 5 \
+--per_device_train_batch_size 4 \
+--per_device_eval_batch_size 4 \
+--gradient_accumulation_steps 4 \
+--evaluation_strategy "no" \
+--save_strategy "epoch" \
+--learning_rate 1e-5 \
+--weight_decay 0. \
+--warmup_ratio 0.03 \
+--lr_scheduler_type "cosine" \
+--logging_steps 1 \
+--fsdp "full_shard auto_wrap" \ 
+--fsdp_config tuning/config/fsdp_config.json \
+--include_tokens_per_second \
+--packing False \
+--response_template "\n### Response:" \
+--dataset_text_field "output"
+```
+
+The above is an example. We would need to tune parameters depending on the model size, data size. The above example has been validated on 8 x A100 80GB.
+
+For `GPTBigCode` models, Hugging Face has enabled Flash v2 and one can simply replace the `'LlamaDecoderLayer'` with `'GPTBigCodeBlock'` for proper sharding of the model.
```

Comment:
replace in `tuning/config/fsdp_config.json`? might want to specify where the change happens since I was expecting to see 'LlamaDecoderLayer' in the script above.